### PR TITLE
Testing improvements

### PR DESCRIPTION
**Pull request by martijnwalraven**
This PR builds on the use of TypeScript project references to improve the testing workflow. 

Previously, test files were ignored by our `tsconfig.json` files, which meant that they weren't properly type checked. Opening a test file in VS Code relied on the automatic creation of an implicit project with default settings by the TypeScript language server.

Adding explicit per-package `tsconfig.json` files for testing, extending a shared `tsconfig.test.base.json`, allows for proper type checking both inside the editor and as part of the testing workflow. For now, for the project to be detected by the TypeScript language server, the `tsconfig.json` files have been nested under `__tests__`, which allows VS Code to find the right project when opening a test file and provide appropriate diagnostics right away.

We've also switched to per-package Jest projects, with a common base config and a root config in `package.json` that lists all projects. Running `jest` (or `npm test`) from the root will run tests in all projects. Every Jest project will use the corresponding TypeScript test project.

The Jest configuration has also been changed to rely on a `moduleNameMapper` to resolve imports of local packages to their source, instead of the compiler output. That means running tests doesn't require separate compilation, and `jest --watch` will always pick up the latest changes. Compiled files are cached by Jest, so runs are still pretty performant. (There is a `test:clean` run command now to clean the Jest cache, in case of issues.)

To get diagnostics across projects (including for files that haven't been opened), you can run a default build task that runs a TypeScript watch in incremental compilation mode and reports diagnostics in the VS Code problems panel.

Because test files have a lot of typing issues that need to be fixed, they are not currently included in the watch. And we also set `diagnostics: false` in `jest.config.base.js`, to avoid failing tests because of typing issues. Once we've fixed these however, we can turn this back on to including typing as part of testing.

This also temporarily updates the dependency to `typescript@next` to fix issues with watching and incremental compilation, and to allow for some typing improvements.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -3422,7 +3540,7 @@
         },
         "readable-stream": {
           "version": "2.3.6",
-          "resolved": "https://registry.npmjs.org/readable-stream/-/readable-stream-2.3.6.tgz",
+          "resolved": "http://registry.npmjs.org/readable-stream/-/readable-stream-2.3.6.tgz",
```
Comment (abernix) [inline package-lock.json]: Note to self.  change to `https`.

File: package-lock.json [hunk 2]
```diff
@@ -4427,41 +4496,13 @@
...
-            "ansi-regex": "^3.0.0",
-            "ansi-styles": "^3.2.0"
-          }
-        }
       }
     },
     "express": {
       "version": "4.16.3",
-      "resolved": "https://registry.npmjs.org/express/-/express-4.16.3.tgz",
+      "resolved": "http://registry.npmjs.org/express/-/express-4.16.3.tgz",
```
Comment (abernix) [inline package-lock.json]: These things drive me nuts, and I don't understand why.

File: package-lock.json [hunk 3]
```diff
@@ -13034,7 +12860,7 @@
     },
     "yargs": {
       "version": "11.1.0",
-      "resolved": "https://registry.npmjs.org/yargs/-/yargs-11.1.0.tgz",
+      "resolved": "http://registry.npmjs.org/yargs/-/yargs-11.1.0.tgz",
```
Comment (abernix) [inline package-lock.json]: I'll fix these `http` things in a follow-up commit.  I have a slight understanding of why they're happening, but they're annoying.

File: packages/apollo-cache-control/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const config = require('../../jest.config.base');
+
+module.exports = Object.assign({}, config);
```
Comment (abernix) [inline packages/apollo-cache-control/jest.config.js]: Slight preference for `Object.assign(Object.create, config)` here (and in general) to create an object that doesn't extend `Object.prototype`, just because of the potential for conflicting `prototype` properties. (The type of thing that surfaces as a very bizarre edge case).

File: packages/apollo-server-cache-memcached/src/__tests__/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "../../../../tsconfig.test.base",
+  "include": ["**/*"],
+  "references": [
+    { "path": "../../tsconfig.json" },
```
Comment (abernix) [inline packages/apollo-server-cache-memcached/src/__tests__/tsconfig.json]: Just out of curiosity, the explicit `tsconfig.json` files aren't necessary, right? If they were left off it would work the same?  (Though it seems to make sense to be explicit since we occasionally use `tsconfig.test.json`?)

File: packages/apollo-server-cloudflare/tsconfig.test.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "../../tsconfig.test.base",
+  "include": ["src/**/__tests__", "src/**/__mocks__"],
+  "references": [
+    { "path": "./tsconfig.json" },
+    { "path": "../apollo-server-integration-testsuite" }
```
Comment (abernix) [inline packages/apollo-server-cloudflare/tsconfig.test.json]: Still not quite understanding the occasional explicit reference to `tsconfig.json` and occasionally referencing the root path (which I believe then implies `tsconfig.json`?

File: packages/apollo-server-express/jest.config.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+const config = require('../../jest.config.base');
+
+module.exports = Object.assign({}, config);
```
Comment (abernix) [inline packages/apollo-server-express/jest.config.js]: Same comment on this as the earlier one. (plus the other `jest.config.js`'s I didn't comment on)

File: packages/apollo-tracing/src/__tests__/index.ts [hunk 8]
```diff
@@ -1,6 +0,0 @@
-import { TracingExtension } from '../index';
```
Comment (abernix) [inline packages/apollo-tracing/src/__tests__/index.ts]: Glad we're removing these "tests" that are not doing anything at all.  They're not even `xit`!

File: tsconfig.test.json [hunk 9]
```diff
@@ -1,6 +1,22 @@
...
+    { "path": "./packages/apollo-server/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-cache-memcached/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-cache-redis/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-caching/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-core/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-express/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-hapi/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-koa/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-lambda/src/__tests__/tsconfig.json" },
+    { "path": "./packages/apollo-server-micro/src/__tests__/tsconfig.json" },
```
Comment (abernix) [inline tsconfig.test.json]: These _could_ be without `tsconfig.json`, right?

File: jest.config.base.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    testMatch: null,
+    testRegex: "/__tests__/.*\\.test\\.(js|ts)$",
+    testPathIgnorePatterns: [
+      "/node_modules/",
+      "/dist/"
+    ],
+    moduleFileExtensions: [...defaults.moduleFileExtensions, "ts", "tsx"],
+    moduleNameMapper: {
+      '^__mocks__/(.*)$': '<rootDir>/../../__mocks__/$1',
+      '^(?!apollo-server-env|apollo-engine-reporting-protobuf)(apollo-(?:server|datasource|cache-control|tracing|engine)[^/]*|graphql-extensions)(?:/dist)?((?:/.*)|$)': '<rootDir>/../../packages/$1/src$2'
```
Comment (abernix) [inline jest.config.base.js]: This looks right to me — and I really like this approach since, as you said, it allows more correct mapping to the true source.  I assume the `/`s don't need to be escaped because they're passed to `new RegExp("...")`).  So I would probably say:

1. While re-using the `apollo-` prefix is nice, it makes refactoring a bit less clear.  For example, if I was changing the name of `apollo-cache-control` (or creating a new module based off it) a global grep on this repository for `apollo-cache-control` wouldn't bring me to this.  If we explicitly put `apollo-cache-control`, it would.  Though, I know that would make this very long.  I'm not sure what the best solution here is.
2. Since this is a `.js` file, maybe we could add a comment above about the intention here?  Just took me a bit to digest exactly the intion (not `apollo-server-env`, but some specific other packages was, and your explanation might get lost in this PR.
3. `engine` seems like it should be explicitly `engine-reporting` to not inadvertently match something else?

File: packages/apollo-cache-control/jest.config.js (hunk 4 above)
Comment (martijnwalraven) [reply packages/apollo-cache-control/jest.config.js]: Ah, yes, that makes sense. I'd prefer to switch to object spreads when we drop support for Node 6 however. Not sure how those deal with `Object.prototype`.

File: packages/apollo-server-cache-memcached/src/__tests__/tsconfig.json (hunk 5 above)
Comment (martijnwalraven) [reply packages/apollo-server-cache-memcached/src/__tests__/tsconfig.json]: Yes, that's true. I originally used `tsconfig.build.json` at the package-level as well, so we could have a `tsconfig.json` with references to both the build and test project for a package. That didn't work however, because the TypeScript language server (and thus VS Code) doesn't currently follow `references` when looking up a project config for a file.

File: packages/apollo-server-cloudflare/tsconfig.test.json (hunk 6 above)
Comment (martijnwalraven) [reply packages/apollo-server-cloudflare/tsconfig.test.json]: Yep, it probably makes sense to be consistent here and drop it everywhere.

File: tsconfig.test.json (hunk 9 above)
Comment (martijnwalraven) [reply tsconfig.test.json]: Yep :)
